src/GafferImage/ColorProcessor.cpp

```diff
@@ -141,22 +178,66 @@ void ColorProcessor::hash( const Gaffer::ValuePlug *output, const Gaffer::Contex
 {
 	ImageProcessor::hash( output, context, h );
 
-	if( output == colorDataPlug() )
+	if( output == colorProcessorPlug() )
+	{
+		hashColorProcessor( context, h );
+	}
+	else if( output == colorDataPlug() )
 	{
-		hashColorData( context, h );
+		ConstStringVectorDataPtr channelNamesData;
+		bool unpremult;
+		{
+			ImagePlug::GlobalScope globalScope( context );
+			channelNamesData = inPlug()->channelNamesPlug()->getValue();
+			unpremult = processUnpremultipliedPlug()->getValue();
+			colorProcessorPlug()->hash( h );
+		}
+		const vector<string> &channelNames = channelNamesData->readable();
+
+		const string &layerName = context->get<string>( g_layerNameKey );
+
+		ImagePlug::ChannelDataScope channelDataScope( context );
+		for( const auto &baseName : { "R", "G", "B" } )
+		{
+			string channelName = ImageAlgo::channelName( layerName, baseName );
+			if( ImageAlgo::channelExists( channelNames, channelName ) )
+			{
+				channelDataScope.setChannelName( &channelName );
+				inPlug()->channelDataPlug()->hash( h );
+			}
+			else
+			{
+				ImagePlug::blackTile()->hash( h );
+			}
+		}
+
+		if( unpremult && ImageAlgo::channelExists( channelNames, ImageAlgo::channelNameA ) )
+		{
+			channelDataScope.setChannelName( &ImageAlgo::channelNameA );
+			inPlug()->channelDataPlug()->hash( h );
+		}
 	}
 }
 
 void ColorProcessor::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
 {
-	if( output == colorDataPlug() )
+	if( output == colorProcessorPlug() )
+	{
+		ColorProcessorDataPtr data = new ColorProcessorData;
+		data->colorProcessor = colorProcessor( context );
+		static_cast<ObjectPlug *>( output )->setValue( data );
+		return;
+	}
+	else if( output == colorDataPlug() )
 	{
 		ConstStringVectorDataPtr channelNamesData;
+		ConstColorProcessorDataPtr colorProcessorData;
 		bool unpremult;
 		{
 			ImagePlug::GlobalScope globalScope( context );
 			channelNamesData = inPlug()->channelNamesPlug()->getValue();
 			unpremult = processUnpremultipliedPlug()->getValue();
+			colorProcessorData = boost::static_pointer_cast<const ColorProcessorData>( colorProcessorPlug()->getValue() );
```

Comment:
Should we check if colorProcessorData->colorProcessor is null here before we call it below? Because of the hash pass-through below, we will usually be hitting an upstream cached value, but isn't there the possibility that this compute runs with a null colorProcessor?